netperf_tool/src/server_node.cpp

```diff
@@ -0,0 +1,220 @@
+      "using timestamps added to the message to calculate latency");
+    rclcpp::Time pub_timestamp = msg.timestamp;
+    latency = std::chrono::nanoseconds{(rec_timestamp - pub_timestamp).nanoseconds()};
+  }
+  bool emplaced{false};
+  auto string_gid = stringify_gid(rmw_msg_info.publisher_gid);
+  {
+    // Store message info
+    std::lock_guard guard{pub_gid_to_collected_info_map_mutex_};
+    RCLCPP_INFO_STREAM_ONCE(
+      this->get_logger(),
+      "Added one message with gid: " << string_gid);
+    auto it_emplaced_pair = pub_gid_to_collected_info_map_.try_emplace(string_gid);
+    auto & collected_info = it_emplaced_pair.first->second;
+    ServerMessageInfo info;
+    info.id = msg.id;
+    info.latency = latency;
+    info.serialized_size = received_bytes;
+    info.reception_time = rec_timestamp;
+    collected_info.message_infos.emplace_back(info);
+    emplaced = it_emplaced_pair.second;
+  }
+  if (emplaced) {
+    // first time
+    std::lock_guard guard{notify_mutex_};
+    new_clients_.emplace_back(string_gid);
+  }
+  notify_cv_.notify_one();
+}
+
+void
+ServerNode::handle_get_results_request(
+  netperf_tool_interfaces::srv::GetResults::Request::SharedPtr req,
+  netperf_tool_interfaces::srv::GetResults::Response::SharedPtr rep)
+{
+  ServerCollectedInfo collected_info;
+  {
+    std::lock_guard guard{pub_gid_to_collected_info_map_mutex_};
+    auto it = pub_gid_to_collected_info_map_.find(req->publisher_gid);
+    if (it == pub_gid_to_collected_info_map_.end() || 0u == it->second.message_infos.size()) {
+      return;
+    }
+    collected_info = std::move(it->second);
+    // remove data from that publisher when the results are requested
+    pub_gid_to_collected_info_map_.erase(it);
+  }
+  // calculate latency statistics
+  auto latency_avg_ms = std::accumulate(
+    collected_info.message_infos.begin(),
+    collected_info.message_infos.end(),
+    0.0,
+    [n_items = static_cast<double>(collected_info.message_infos.size())](auto lhs, auto rhs) {
+      return lhs + static_cast<double>(rhs.latency.count()) / 1e6 / n_items;
+    });
+  auto latency_var_ms2 = std::accumulate(
+    collected_info.message_infos.begin(),
+    collected_info.message_infos.end(),
+    0.0,
+    [latency_avg_ms, div = static_cast<double>(collected_info.message_infos.size() - 1)](
+      auto lhs, auto rhs)
+    {
+      auto diff = static_cast<double>(rhs.latency.count()) / 1e6 - latency_avg_ms;
+      return lhs + diff * diff / div;
+    });
+  auto latency_min_max_its = std::minmax_element(
+    collected_info.message_infos.begin(),
+    collected_info.message_infos.end(),
+    [](auto lhs, auto rhs) {
+      return lhs.latency < rhs.latency;
+    });
+  rep->latency_avg_ms = latency_avg_ms;
+  rep->latency_stdev_ms = std::sqrt(latency_var_ms2);
+  rep->latency_min_ms = static_cast<double>(latency_min_max_its.first->latency.count()) / 1e6;
+  rep->latency_max_ms = static_cast<double>(latency_min_max_its.second->latency.count()) / 1e6;
```

Comment:
For the NS to MS conversions in this section, it might be more readable if you either use something like:
```cpp
using ms = std::chrono::duration<double, std::milli>;

. . .

return lhs + std::chrono::duration_cast<ms>(rhs.latency).count() / n_items;
```

Or if there's a gotcha that I'm missing with that approach, just a `constexptr auto NS_TO_MS {1/1e6};`?

Reply:
Done in https://github.com/ros2/netperf/pull/1/commits/257523aa13b6b81126a5a804648b054b19e39939.